### ISSUE TITLE
`used-before-assignment` false negative after conditionals

### ISSUE THREAD

**Issue by eddieViscosity** [pylint-dev/pylint#1727]
Running pylint on this simple example does not detect that x can be used before assignment.  This is a very common programming mistake and should be caught.
```python
def a_function(d):
    if d:
        x=1
    return x
```

**Comment by kylebebak** [comment]
Just here to second this.

**Comment by nickdrozd** [comment]
Ran into this at work today!

**Comment by nickdrozd** [comment]
This kind of problem would be avoided entirely if the variable were set [with a ternary expression](https://github.com/PyCQA/pylint/pull/4871)!

**Comment by jacobtylerwalls** [comment]
Duplicate of #267 but some good discussion here, so maybe we can choose one issue to keep open.

We can possibly explore handling this after the precedent in #5402 of storing `consumed_uncertain` (name to be finalized before merge).

**Comment by Pierre-Sassoulas** [comment]
There's some more example in #267 or #1807 for functional tests.

**Comment by jacobtylerwalls** [comment]
The reason this is not a duplicate of #1727 is that we could issue `possibly-used-before-assignment` here for the ambiguity. Whereas in #1727 we could be bolder, detect "always False", and issue the garden variety `used-before-assignment` (with a CONTROL_FLOW confidence level).

**Comment by Pierre-Sassoulas** [comment]
@jacobtylerwalls I think you wanted to post this message in another issue because this one is #1727 :D

**Comment by alonme** [comment]
@jacobtylerwalls @Pierre-Sassoulas - i would like to help implement this,
Any pointers?

**Comment by jacobtylerwalls** [comment]
Very cool!

I'd familiarize yourself with this method and how it is used:

https://github.com/pylint-dev/pylint/blob/7eaac577d7e6e6ab3d299c12d22c4b9b8723ddfa/pylint/checkers/variables.py#L687-690

Essentially, thanks to hard work by @zenlyj, we have a pretty clean interface for emitting used-before-assignment or not based on whether or not we care about a certain if/else pattern. We've been ratcheting up the messages slowly. First we just started emitting used-before-assignment only when we can tell that all possible values of the test are False (and therefore the if is always skipped).

So, this issue is about being even bolder, and emitting _something_ (either used-before-assignment or a new message possibly-used-before-assignment) regardless of the inferred possible values of the test (see the `all_inferred` variable).

It might be nice to get a draft PR going without defining a separate message just to see how noisy it would be. Then we can evaluate whether to create a second message for it. Does that make sense?

**Comment by jacobtylerwalls** [comment]
> Do we want to create a new checker called possibly-used-before-assignment checker for this? Similar to how there is possibly-unused-variable. If we can't infer d we will never be able to emit used-before-assignment without any doubt.

Ah -- I'm thinking we shouldn't have a second message type, as it would blur the boundary between message type and confidence. We should just be _really_ specific about the confidence, including using `INFERENCE_FAILURE` for this case. And then mention in the release notes that if you find this noisy, you need to experiment with adjusting confidence levels.

**Comment by Pierre-Sassoulas** [comment]
Thank you for your interest in this issue :) Jacob worked a lot on it and is definitely the code owner on this message, I agree with what was said do not hesitate to open a draft PR and use primers in our CI as a dev tool.
